src/main/java/com/amazon/opendistroforelasticsearch/sql/rewriter/nestedfield/NestedFieldProjection.java

```diff
@@ -123,23 +135,64 @@ private void buildInnerHit(List<String> fieldNames, NestedQueryBuilder query) {
      * Why linear search? Because NestedQueryBuilder hides "path" field from any access.
      * Assumption: collected NestedQueryBuilder list should be very small or mostly only one.
      */
-    private NestedQueryBuilder findNestedQueryWithSamePath(List<NestedQueryBuilder> nestedQueries, String path) {
+    private NestedQueryBuilder findNestedQueryWithSamePath(
+            List<NestedQueryBuilder> nestedQueries, String path, List<String> fieldNames) {
+
         return nestedQueries.stream().
-                             filter(query -> isSamePath(path, query)).
-                             findAny().
-                             orElseGet(createEmptyNestedQuery(path));
+                filter(query -> isSamePath(path, query, fieldNames)).
+                findAny().
+                orElseGet(createNestedQuery(path, fieldNames));
+    }
+
+    /** Check against NestedQuery of WHERE condition */
+    private boolean isSamePath(String path, NestedQueryBuilder query, List<String> fieldNames) {
+        boolean isSamePath = nestedQuery(path, query.query(), query.scoreMode()).equals(query);
+        if (isSamePath) {
```

Comment:
This is misleading - you function mutates the state, that is not clear for consumer.
Instead of this, remove condition here and rewrite function above as

```
 return nestedQueries.stream()
                             .filter(query -> isSamePath(path, query, fieldNames)).
                             .findAny()
                             .map(x->setLastValues(path,fieldNames))
                             orElseGet(createEmptyNestedQuery(path));	                orElseGet(createNestedQuery(path, fieldNames));
```

This findAny is also quite suspicious

Reply:
or use peek() if you want to modify the collection

Reply:
will use peek here. Thanks.